src/wiringx.c

```diff
@@ -103,6 +86,142 @@ static struct spi_t spi[2] = {
 	} while (0)
 #endif
 
+#ifndef __FreeBSD__
+/* SPI Bus Parameters */
+struct spi_t {
+	uint8_t mode;
+	uint8_t bits_per_word;
+	uint16_t delay;
+	uint32_t speed;
+	int fd;
+} spi_t;
+
+/* SPI states list implementation */
+
+struct spi_id_t {
+	uint8_t device;
+	uint8_t channel;
+};
+
+struct spi_list_t {
+	struct spi_id_t id;
+	struct spi_t data;
+};
+
+struct spi_list_t *spilist = NULL;
+int nrspilist = 0;
+
+int spi_new(struct spi_id_t id) {
+	int entry = -1;
+	void *tmp = NULL;
+	int i = 0;
+
+	// loop over existing states to detect potential duplicates
+	for(i = 0; i < nrspilist; i++) {
+		if(memcmp(&spilist[i].id, &id, sizeof(struct spi_id_t)) == 0) {
+			entry = i;
+
+			/* clear existing entry */
+			if(spilist[entry].data.fd > 0) {
+				close(spilist[entry].data.fd);
+			}
+
+			/* return index of existing entry */
+			return entry;
+		}
+	}
+
+	/* reserve space for new entry (+1) */
+	entry = nrspilist++;
+	tmp = realloc(spilist, sizeof(struct spi_list_t)*nrspilist);
+	if(tmp == NULL) {
+		fprintf(stderr, "out of memory\n");
+		exit(EXIT_FAILURE);
+	}
+	spilist = tmp;
+	spilist[entry].id = id;
+	memset(&spilist[entry].data, 0, sizeof(struct spi_t));
+
+	/* return index of new entry */
+	return entry;
+}
+
+void spi_clear() {
+	struct spi_list_t *entry = NULL;
+	int i = 0;
+
+	/* loop over all entries */
+	for(i = 0; i < nrspilist; i++) {
+		entry = &spilist[i];
+
+		/* clear */
+		if(entry->data.fd > 0) {
+			close(entry->data.fd);
+		}
+	}
+
+	/* free storage */
+	free(spilist);
+	spilist = NULL;
+	nrspilist = 0;
+}
+
+/* test-case for list */
+int spi_test() {
```

Comment:
`static`

Reply:
Do you want to keep this in the final code?

Reply:
Yes I do. Only question is style-wise:
Should test-cases for data-structures (not API), like here, live in the source file that they test, or elsewhere?
This has influence on wether we can declare things as static or not.

I'd suggest to have one (component)_test() function in every source file that has tests.
And then another source file that would call all _test functions, and will produce a standalon test binary.